File: Year-2023/LeetCode-August-Challenge/MinimizetheMaximumDifferenceofPairs.cpp
#include<bits/stdc++.h>
using namespace std;

class Solution {
private:
    bool isPair(vector<int> &nums, int mid, int p){
        int count = 0;
        for(int i = 0; i < nums.size() - 1 && count < p; ++i){
            if(nums[i + 1] - nums[i] <= mid){
                count++;
                i++;
            }
        }
        return count >= p;
    }
public:
    int minimizeMax(vector<int>& nums, int p) {
        sort(nums.begin(), nums.end());
        int low = 0, high = nums.back() - nums.front();

        while(low < high){
            int mid = low + (high - low)/2;
            if(isPair(nums, mid, p)) high = mid;
            else low = mid + 1;
        }

        return low;
    }
};

int main(){
    vector<int> nums = {10,1,2,7,1,3}; int p = 2;
    Solution *s = new Solution();
    cout<<s->minimizeMax(nums, p)<<endl;
    return 0;
}

File: Year-2023/LeetCode-August-Challenge/MinimizetheMaximumDifferenceofPairs.js
var minimizeMax = function (nums, p) {

    nums.sort((a, b) => a - b);

    const isPair = (nums, mid, p) => {
        let count = 0;
        for (let i = 0; i < nums.length - 1 && count < p; ++i) {
            if (nums[i + 1] - nums[i] <= mid) {
                count++;
                i++;
            }
        }

        return count >= p;
    }

    let low = 0, high = nums[nums.length - 1] - nums[0];

    while (low < high) {
        let mid = low + Math.trunc((high - low) / 2);
        if (isPair(nums, mid, p)) high = mid;
        else low = mid + 1;
    }

    return low;
};

let nums = [10,1,2,7,1,3], p = 2;
console.log(minimizeMax(nums, p));